peregrine/iqgen/bits/doppler_const.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (C) 2016 Swift Navigation Inc.
+#
+# This source is subject to the license found in the file 'LICENSE' which must
+# be be distributed together with this source. All other rights reserved.
+#
+# THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+# EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+
+"""
+The :mod:`peregrine.iqgen.doppler_const` module contains classes and functions
+related to generation of zero-doppler signals.
+
+"""
+
+import scipy.constants
+from peregrine.iqgen import if_iface
+
+class Doppler(object):
+
+  NAME = "ConstDoppler"
+
+  '''
+  Doppler control for a signal source that moves with a constant speed.
+  '''
+  def __init__(self, distance0_m, speed_mps):
+    '''
+    Constructs doppler control object for linear movement.
+
+    Parameters
+    ----------
+    distance0_m : float
+      Initial distance to satellite at time 0 in meters for computing signal
+      phase and delay.
+    speed_mps : float
+      Speed of satellite in meters per second.
+    '''
+    super(Doppler, self).__init__()
+    self.distance0_m = distance0_m
+    self.speed_mps = speed_mps
+    self.c1 = self.distance0_m / (scipy.constants.c + self.speed_mps)
+    self.c2 = self.speed_mps / (scipy.constants.c + self.speed_mps)
+
+  def computeDistanceM(self, svTime_s):
+    '''
+    Computes distance to satellite in meters.
+
+    Parameters
+    ----------
+    svTime_s : float
+      Time in seconds at which distance is computed. Please note that  is not
+      a time of the observer.
+
+    Returns
+    -------
+    float
+      Distance to satellite in meters.
+    '''
+    return self.distance0_m + self.speed_mps * svTime_s
+
+  def computeSpeedMps(self, svTime_s):
+    '''
+    Computes speed of satellite in meters per second.
+
+    Parameters
+    ----------
+    svTime_s : float
+      Time in seconds at which speed is computed. Please note that  is not
+      a time of the observer.
+
+    Returns
+    -------
+    float
+      Speed of satellite in meters per second.
+    '''
+    return self.speed_mps
+
+  def computeSvTimeS(self, userTime_s):
+    '''
+    Computes SV time from a user time.
+
+    The computation is a solution of the formulae:
+    T_user = T_sv + D(T_sv)/c, for D(t) = D_0 + V*T_sv
+
+    Parameters
+    ----------
+    userTime_s : float
+      Observer's time in seconds
+
+    Returns
+    -------
+    float
+      Satellite vehicle's time at which signal has been generated.
+    '''
+
+    tau_s = self.c1 + self.c2 * userTime_s
+    svTime_s = userTime_s - tau_s
+
+    return svTime_s
+
+  def computeBatch(self,
```

Comment:
1. We add the noise if the appropriate parameter is specified. It is in generate.py.
2. Yes, planFrequency_hz is a carrier frequency. I'll update the implementation.
3. Tracking time internally improves the computation in some cases (not too much though), but makes it difficult to use the class for parallel computations (future extension). In most cases the delay for this computation is not that visible on overall performance.